tests/unit/nupic/algorithms/knn_classifier_test.py

```diff
@@ -0,0 +1,146 @@
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([1, 3, 7, 11, 13, 17, 19, 23, 29], dtype=np.int32)
+    b = np.array([2, 4, 8, 12, 14, 18, 20, 28, 30], dtype=np.int32)
+
+    numPatterns = classifier.learn(a, 0, isSparse=dimensionality)
+    self.assertEquals(numPatterns, 1)
+
+    numPatterns = classifier.learn(b, 1, isSparse=dimensionality)
+    self.assertEquals(numPatterns, 2)
+
+    denseA = np.zeros(dimensionality)
+    denseA[a] = 1.0
+    cat, _, _, _ = classifier.infer(denseA)
+    self.assertEquals(cat, 0)
+
+    denseB = np.zeros(dimensionality)
+    denseB[b] = 1.0
+    cat, _, _, _ = classifier.infer(denseB)
+    self.assertEquals(cat, 1)
+
+
+  def testOverlapDistanceMethod_BadSparsity(self):
+    """Sparsity (input dimensionality) less than input array"""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([1, 3, 7, 11, 13, 17, 19, 23, 29], dtype=np.int32)
+
+    # Learn with incorrect dimensionality, less than some bits (23, 29)
+    with self.assertRaises(RuntimeError):
+      classifier.learn(a, 0, isSparse=20)
+
+
+  def testOverlapDistanceMethod_InconsistentDimensionality(self):
+    """Inconsistent sparsity (input dimensionality)"""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([1, 3, 7, 11, 13, 17, 19, 23, 29], dtype=np.int32)
+
+    # Learn with incorrect dimensionality, greater than largest ON bit, but
+    # inconsistent when inferring
+    numPatterns = classifier.learn(a, 0, isSparse=31)
+    self.assertEquals(numPatterns, 1)
+
+    denseA = np.zeros(dimensionality)
+    denseA[a] = 1.0
+
+    cat, _, _, _ = classifier.infer(denseA)
+    self.assertEquals(cat, 0)
+
+
+
+  def testOverlapDistanceMethod_StandardUnsorted(self):
+    """If sparse representation indices are unsorted expect error."""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([29, 3, 7, 11, 13, 17, 19, 23, 1], dtype=np.int32)
+    b = np.array([2, 4, 20, 12, 14, 18, 8, 28, 30], dtype=np.int32)
+
+    with self.assertRaises(RuntimeError):
+      classifier.learn(a, 0, isSparse=dimensionality)
+
+    with self.assertRaises(RuntimeError):
+      classifier.learn(b, 1, isSparse=dimensionality)
+
+
+  def testOverlapDistanceMethod_EmptyArray(self):
+    """Tests case where pattern has no ON bits"""
+    params = {"distanceMethod": "rawOverlap"}
+    classifier = KNNClassifier(**params)
+
+    dimensionality = 40
+    a = np.array([], dtype=np.int32)
+
+    numPatterns = classifier.learn(a, 0, isSparse=dimensionality)
+    self.assertEquals(numPatterns, 1)
+
+    denseA = np.zeros(dimensionality)
+    denseA[a] = 1.0
+    cat, _, _, _ = classifier.infer(denseA)
+    self.assertEquals(cat, 0)
+
+
+  # def testOverlapDistanceMethod_ClassifySparse(self):
```

Comment:
If you plan to merge this before it works then uncomment it and use the `@unittest.skip("Explanation for why it's skipped with link to GH issue")` decorator to denote this.